### PR TITLE
feat!: rework input argument flags

### PR DESCRIPTION
**Pull request by dundargoc**
Remove `--dir` flag as it's incomprehensible and doesn't seem to work.
Adjust `--outdir` to create the directory name if it doesn't exist.
Remove `--file` flag and make the filename a required input.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -120,11 +40,20 @@ fn main() -> Result<()> {
...
-        let file_name = path.file_name().unwrap();
+        let path = Path::new(&args.outdir);
+        if path.is_file() {
+            return Err(anyhow::anyhow!(
+                "Can't create directory: a file with the same name already exists!"
+            ));
+        } else if path.is_dir() {
+        } else {
+            std::fs::create_dir(&args.outdir)?;
+        }
```
Comment (seandewar) [inline src/main.rs]:
```suggestion
        if path
            .try_exists()
            .context("Couldn't determine if file already exists")?
        {
            ensure!(
                path.is_dir(),
                "Can't create directory: a file with the same name already exists!"
            );
        } else {
            std::fs::create_dir(&args.outdir)?;
        }
```

Not entirely sure if this handles symlinks and other types properly. A better solution may be to check for existence instead.

Anyhow's `ensure!(x, "foo")` can be used to simplify the `if !x { return anyhow!("foo"); }` pattern.
`context()` can be used to add additional context to errors before bubbling them up (for better error messages).

File: src/main.rs [hunk 2]
```diff
@@ -1,109 +1,29 @@
...
 
 #[derive(Parser, Debug)]
 #[command(author, version, about, long_about = None)]
 struct Args {
-    /// The directory for generating paths
-    #[arg(short, long)]
-    dir: Option<String>,
-
     #[arg(long)]
     file: Option<String>,
```
Comment (seandewar) [inline src/main.rs]:
```suggestion
    file: String,
```

No need to make this optional or a `--file` argument any more; we can tell clap to verify that it's there.

File: src/main.rs [hunk 3]
```diff
@@ -1,109 +1,29 @@
...
-        gen_directory(&dir_src, &dir_gen, OsStr::new("autoload"))?;
-        return Ok(());
-    } else if let Some(file) = args.file {
-        let path = Path::new(&file);
-        if !path.is_file() {
+    if let Some(file) = args.file {
+        let filepath = Path::new(&file);
+        if !filepath.is_file() {
             return Err(anyhow::anyhow!("not a valid file"));
         }
```
Comment (seandewar) [inline src/main.rs]:
```suggestion
    let filepath = Path::new(&args.file);
    ensure!(filepath.is_file(), "not a valid file");
```

`args.file` is no longer optional.
Anyhow's `ensure!(x, "foo")` can be used to simplify the `if !x { return anyhow!("foo"); }` pattern.
This also gets rid of an indent :rocket:

File: src/main.rs [hunk 4]
```diff
@@ -138,5 +67,5 @@ fn main() -> Result<()> {
         return Ok(());
     }
 
-    Err(anyhow::anyhow!("Need to pass either --dir or --file"))
+    Err(anyhow::anyhow!("Need to pass --file"))
```
Comment (seandewar) [inline src/main.rs]: This can be yoted with the other changes.

File: src/main.rs [hunk 5]
```diff
@@ -120,11 +40,20 @@ fn main() -> Result<()> {
             }
         };
```
Comment (seandewar) [inline src/main.rs]:
```rust
    let generated = gen::generate(
        &contents,
        gen::ParserOpts {
            mode: gen::ParserMode::Autoload {
                // TODO: Calculate this correctly...
                prefix: "ccomplete".to_string(),
            },
        },
    )
    .unwrap_or_else(|err| {
        println!("  FAILED: {}", err.1);
        err.0
    });
```
This whole `generated` thing might look slightly clearer with `unwrap_or_else()`.

**Review by seandewar — COMMENTED** [review]
Most of this review assumes you `use anyhow::{ensure, Context, Result};`

File: src/main.rs (hunk 1 above)
Comment (seandewar) [reply src/main.rs]:
```rust
    match std::fs::metadata(path) {
        Ok(m) if m.is_dir() => {}
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => std::fs::create_dir(&args.outdir)?,

        Ok(_) => bail!("Can't create directory: a file with the same name already exists!"),
        Err(e) => Err(e).context("Failed to check out_dir")?,
    }
```
try this I guess lol

**Comment by dundargoc** [comment]
Thanks for the help

**Pull request merged**